common/src/main/java/com/wynntils/wynn/model/quests/QuestManager.java

```diff
@@ -133,21 +144,96 @@ public Location getCurrentQuestLocation() {
     }
 
-    protected void setCurrentQuest(QuestInfo questInfo) {
-        currentQuest = questInfo;
-        WynntilsMod.postEvent(new TrackedQuestUpdateEvent(currentQuest));
+    public void clearTrackedQuestFromScoreBoard() {
+        updateTrackedQuest(null);
+    }
+
+    public void updateTrackedQuestFromScoreboard(String name, String nextTask) {
+        // If our quest book has not yet been scanned, we can't update now
+        // but will do after scanning is complete
+        if (updateAfterRescan(name, nextTask)) return;
+
+        Optional<QuestInfo> questInfoOpt;
+        questInfoOpt = getQuestInfoFromName(name);
+        if (questInfoOpt.isEmpty()) {
+            WynntilsMod.warn("Cannot match quest from scoreboard to actual quest: " + name);
+            return;
+        }
+
+        QuestInfo questInfo = questInfoOpt.get();
+        questInfo.setNextTask(nextTask);
+
+        updateTrackedQuest(questInfo);
+    }
+
+    private void updateTrackedQuest(QuestInfo questInfo) {
+        trackedQuest = questInfo;
+        WynntilsMod.postEvent(new TrackedQuestUpdateEvent(trackedQuest));
+    }
+
+    private Optional<QuestInfo> getQuestInfoFromName(String name) {
+        Optional<QuestInfo> questInfoOpt;
+        if (name.startsWith("Mini-Quest - ")) {
+            String shortName = StringUtils.replaceOnce(name, "Mini-Quest - ", "");
+            questInfoOpt = Managers.Quest.miniQuests.stream()
+                    .filter(quest -> quest.getName().equals(shortName))
+                    .findFirst();
+        } else {
+            questInfoOpt = Managers.Quest.quests.stream()
+                    .filter(quest -> quest.getName().equals(name))
+                    .findFirst();
+        }
+        return questInfoOpt;
+    }
+
+    private boolean updateAfterRescan(String name, String nextTask) {
+        if (name.startsWith("Mini-Quest - ")) {
+            if (miniQuests.isEmpty()) {
+                afterRescanTask = nextTask;
+                String shortName = StringUtils.replaceOnce(name, "Mini-Quest - ", "");
+                afterRescanName = shortName;
+                rescanQuestBook(false, true);
+                return true;
+            }
+            return false;
+        } else {
+            if (quests.isEmpty()) {
+                afterRescanTask = nextTask;
+                afterRescanName = name;
+                rescanQuestBook(true, false);
+                return true;
+            }
+            return false;
+        }
     }
 
-    protected void setQuests(List<QuestInfo> newQuests) {
+    protected void updateQuestsFromQuery(List<QuestInfo> newQuests, QuestInfo trackedQuest) {
         quests = newQuests;
+        maybeUpdateTrackedQuest(trackedQuest);
         WynntilsMod.postEvent(new QuestBookReloadedEvent.QuestsReloaded());
     }
 
-    protected void setMiniQuests(List<QuestInfo> newMiniQuests) {
+    protected void updateMiniQuestsFromQuery(List<QuestInfo> newMiniQuests, QuestInfo trackedQuest) {
```

Comment:
private?

Reply:
Same as above. I do whatever you decide on both places.